pageserver/src/tenant.rs

```diff
@@ -1462,13 +1492,119 @@ impl Tenant {
     }
 
     /// Shuts down a timeline's tasks, removes its in-memory structures, and deletes its
-    /// data from disk.
-    ///
-    /// This doesn't currently delete all data from S3, but sets a flag in its
-    /// index_part.json file to mark it as deleted.
-    pub async fn delete_timeline(
+    /// data from both disk and s3.
+    #[instrument(skip(self, timeline))]
+    async fn delete_timeline(
         &self,
         timeline_id: TimelineId,
+        timeline: Arc<Timeline>,
+    ) -> anyhow::Result<()> {
+        {
```

Comment:
This part is just moved out to different function. Should be identical till the end where the only difference is call to `remote_client.delete_all().await?;`

Reply:
Yeah thanks for this one, it was a bit confusing :) Will we be left with a possible leaking s3 files issue here via crash at inopportune time? If so, could we note it here in the beginning? Also link the issue if you created one.

Reply:
I think we shouldnt. The sequence in #4326 is:

1. Schedule layer deletion
2. Upload index_file.json change
3. Crash before actual deletion is done

And here we are not uploading index part anymore. 

The part that I've missed is that we should ignore not found errors during deletion for the case when we resume deletion. Will do

Reply:
Ignoring a delete error is non trivial currently, because we use anyhow all the way down the stack. I think I'll do it as a follow up. Resolving, feel free to reopen

Reply:
Had to deal with it to fix the test anyway, as it turns out s3 delete doesnt return an error when you delete a nonexisting object, so I've just adjusted LocalFs impl to do that. cc @koivunej

Reply:
Added Rust integration test for that

Reply:
> as it turns out s3 delete doesnt return an error when you delete a nonexisting object

Interesting but makes sense for their strong consistency guarantees.